lib/ansible/modules/cloud/amazon/ec2_eip.py

```diff
@@ -278,125 +315,132 @@ def allocate_address(ec2, domain, reuse_existing_ip_allowed):
 
-    raise EIPException("could not find instance" + device_id)
+    raise EIPException("could not find instance " + device_id)
 
 
 def ensure_present(ec2, module, domain, address, private_ip_address, device_id,
-                   reuse_existing_ip_allowed, allow_reassociation, check_mode, isinstance=True):
-    changed = False
+                   reuse_existing_ip_allowed, allow_reassociation, check_mode, is_aninstance=True):
 
     # Return the EIP object since we've been given a public IP
+
     if not address:
         if check_mode:
-            return {'changed': True}
+            changed = True
+            return changed
 
-        address, changed = allocate_address(ec2, domain, reuse_existing_ip_allowed)
+        address, changed = allocate_address(ec2, domain,
+                                            reuse_existing_ip_allowed)
 
     if device_id:
         # Allocate an IP for instance since no public_ip was provided
-        if isinstance:
+        if is_aninstance:
             instance = find_device(ec2, module, device_id)
             if reuse_existing_ip_allowed:
-                if instance.vpc_id and len(instance.vpc_id) > 0 and domain is None:
-                    raise EIPException("You must set 'in_vpc' to true to associate an instance with an existing ip in a vpc")
+                if instance.get('VpcId') and len(instance.get('VpcId')) > 0 and domain is None:
+                    raise EIPException(
+                        "You must set 'in_vpc' to true to associate an instance with an existing ip in a vpc")
             # Associate address object (provided or allocated) with instance
-            assoc_result = associate_ip_and_device(ec2, address, private_ip_address, device_id, allow_reassociation,
-                                                   check_mode)
+            response, changed = associate_ip_and_device(ec2, address, private_ip_address, device_id, allow_reassociation,
+                                                        check_mode)
+
         else:
-            instance = find_device(ec2, module, device_id, isinstance=False)
+            instance = find_device(ec2, module, device_id, is_aninstance=False)
             # Associate address object (provided or allocated) with instance
-            assoc_result = associate_ip_and_device(ec2, address, private_ip_address, device_id, allow_reassociation,
-                                                   check_mode, isinstance=False)
+            response, changed = associate_ip_and_device(ec2, address, private_ip_address, device_id, allow_reassociation,
+                                                        check_mode, is_aninstance=False)
 
-        if instance.vpc_id:
+        if instance.get('VpcId'):
             domain = 'vpc'
 
-        changed = changed or assoc_result['changed']
-
-    return {'changed': changed, 'public_ip': address.public_ip, 'allocation_id': address.allocation_id}
+    return response, changed
```

Comment:
This response is insufficient.  The response from associate_address does not always contain a public_ip, for example.  You will need to grab that from address.